torch/_jit_internal.py

```diff
@@ -1154,8 +1172,19 @@ def persistent_id(self, obj):
         if isinstance(obj, torch.Tensor):
             self.tensors.append(obj)
             return ""
-        else:
-            return None
+        # Since we just want to extract tensors, we don't mind if an object is
+        # unpicklable if it doesn't contain tensors, as we can just ignore/skip
+        # it. To play it safe, we only do so for common objects that we're sure
+        # don't contain tensors. Feel free to add new types here. Note also that
+        # even if a type isn't listed here this won't block users, since thet
+        # can just add a __getstate__ or __reduce__ method to their class.
+        if isinstance(obj, LockType):
+            return ""
+        # Futures and RRefs don't technically contain a value, they just offer
+        # the means to access a value.
+        if isinstance(obj, CFuture) or is_rref_instance(obj):
```

Comment:
Trying to understand this. We don't care about Future and RRef even though they could contain tensors. Is it because we assume Future and RRef themselves should have already handle their Tensors properly?

Reply:
I think at the end of the day this boils down to convention. [IValue::visit](https://github.com/pytorch/pytorch/blob/d3a8505ee1fbb79ba75962eea7891a0aaafc0b0a/aten/src/ATen/core/ivalue.cpp#L183-L184) and [IValue::getSubValues](https://github.com/pytorch/pytorch/blob/d3a8505ee1fbb79ba75962eea7891a0aaafc0b0a/aten/src/ATen/core/ivalue.cpp#L236-L241) don't "enter" Futures when they inspect them. Similarly if you take the future returned by `collectAll` and wait on it, it will _not_ automatically synchronize with the CUDA events of the "internal" futures: that must be done separately, manually.

Given all that it feels natural and consistent here to also not "enter" Futures and RRefs when extracting tensors.